docs/pai-protocol.yaml

```diff
@@ -0,0 +1,217 @@
+    name: String, required
+    type: String, required # Component type. Must be one of the following: data, script, dockerimage, or output. Prerequisites.type cannot be "job"
+    version: String, optional # Component version. Default is latest
+    contributor: String, optional
+    description: String, optional
+    auth: Object, optional # Only available when the type is dockerimage
+      username: String, optional
+      passward: String, optional
+      registryuri: String, optional
+    uri: String or list, required # Only when the type is data can the uri be a list
+
+parameters: # Optional, can be omitted. If specified, the whole object can be referenced as `$parameters`
+  <param1>: value1Type # Specify name and value of all the referencable parameters that will be used in the whole job template.
+  <param2>: value2Type # Can be referenced by `<% $parameters.param1 %>`, `<% $parameters.param2 %>`
+
+jobRetryCount: Integer, optional # Default is 0
+taskRoles:
+  <name>: String, required  # Name of the taskRole
+   protocol_version: String, optional # Protocol version, default is 2
+    instances: Integer, optional # Default is 1, instances of a taskRole
+    completion:
+      minFailedInstances: Integer or null, optional # Default 1
+      minSucceededInstances: Integer or null, optional # Default null
+    taskRetryCount: Integer, optional # Default is 0
+    dockerImage: String, required # Should reference to a dockerimage defined in prerequisites
+    data: String, optional # Should reference to data defined in prerequisites, the whole data object in this taskRole can be referenced as `$data`
+    output: String, optional # Should reference to output defined in prerequisites, the whole output object in this taskRole can be referenced as `$output`
+    script: String, optional # Should reference to script defined in prerequisites, the whole script object in this taskRole can be referenced as `$script`
+    extraContainerOptions:
+      shmMB: Integer, optional # config the /dev/shm in a docker container, https://docs.docker.com/compose/compose-file/#shm_size
+    resourcePerInstance:
+      cpu: Integer, required
+      memoryMB: Integer, required
+      gpu: Integer, required
+      ports:
+        <portLabel1>: Integer, optional, default is 0 # Only for host network
+    commands:
+      - String, required
+
+# to handle that a component may interact with different component differently, user is encouraged to place the codes handling such difference in the "deployments" field.
+# e.g., a job may get input data through wget, hdfc -dfs cp, copy, or just directly read from remote storage. This logic can be placed here.
+# in summary, the deployments field is responsible to make sure the job to run properly in a deployment specific runtime environment.
+# one could have many deployments, but only the first deployment can be activated at runtime. User can choose the deployment at job submission time.
```

Comment:
"but only the first deployment can be activated at runtime." but only one deployment can be activated at runtime (specified in "default")

Reply:
Updated.